docs/architecture/adr-036-blockchain-reactor-refactor.md

```diff
@@ -0,0 +1,406 @@
+
+		case EventNewBlock:
+			if state.PeerMap[event.PeerID] {
+				peerStats = state.PeerMap[event.PeerID]
+				if peerStats.PendingRequest == event.Height {
+					peerStats.PendingRequest = -1
+					state.PendingRequestsNum--
+					if event.CurrentHeight > peerStats.Height {
+						peerStats.Height = event.CurrentHeight
+					}
+					state.Store[event.Height] = BlockInfo{event.Block, event.Commit, event.PeerID}
+					state = verifyBlocks(state) // it can lead to event.PeerID being removed from peer list
+
+					if _, ok := state.PeerMap[event.PeerID]; ok {
+						msg = createBlockRequestMessage(state, event.PeerID, peerStats.Height)
+						if msg != nil {
+							peerStats.PendingRequests = msg.Height
+							state.PendingRequestsNum++
+							timeout = ResponseTimeoutTrigger{msg.PeerID, msg.Height, PeerTimeout}
+						} else if state.PendingRequestsNum == 0 {
+							timeout = TerminationTimeoutTrigger{state.Height, TerminationTimeout}
+						}
+					}
+				} else {
+					error = "Received Block from wrong peer!"
+				}
+			} else {
+				error = "Received Block from unknown peer!"
+			}
+
+			state.PeerMap[event.PeerID] = peerStats
+			return state, msg, timeout, error
+
+		case EventResponseTimeout:
+			if _, ok := state.PeerMap[event.PeerID]; ok {
+				peerStats = state.PeerMap[event.PeerID]
+				if peerStats.PendingRequest == event.Height {
+					add(state.FailedRequests, pendingRequest)
+					delete(state.PeerMap, event.PeerID)
+					state.PendingRequestsNum--
+					if state.PeerMap.isEmpty() || state.PendingRequestsNum == 0 {
+						timeout = TimeoutTrigger{state.Height, TerminationTimeout}
+					}
+				}
+			}
+
+			return state, msg, timeout, error
+
+		case EventTerminationTimeout:
+			if state.Height == event.Height && state.PendingRequestsNum == 0 {
+				state.State = ConsensusMode
+			}
+			return state, msg, timeout, error
+
+		default:
+			error = "Received unknown event type!"
+			return state, msg, timeout, error
+		}
+
+	}
+}
+
+func createBlockResponseMessage(state ControllerState, event BlockRequest) BlockResponseMessage {
+	msg = nil
+	if _, ok := state.PeerMap[event.PeerID]; !ok {
+		peerStats = PeerStats{-1, -1}
+	}
+	if state.Executor.ContainsBlockWithHeight(event.Height) && event.Height > peerStats.Height {
+		peerStats = event.Height
+		msg = BlockResponseMessage{
+			Height:        event.Height,
+			Block:         state.Executor.getBlock(eventHeight),
+			Commit:        state.Executor.getCommit(eventHeight),
+			PeerID:        event.PeerID,
+			CurrentHeight: state.Height - 1,
+		}
+	}
+	state.PeerMap[event.PeerID] = peerStats
+	return msg
+}
+
+func createBlockRequestMessage(state ControllerState, peerID ID, peerHeight int64) BlockRequestMessage {
+	msg = nil
+	blockNumber = -1
+
+	// exist request r in state.FailedRequests such that r <= peerHeight
+	if true {
+		blockNumber = r
+		delete(state.FailedRequests, r)
+	} else if state.MaxRequestPending < peerHeight {
+		state.MaxRequestPending++
```

Comment:
shouldn't we set it to `peerHeight`?

Reply:
also, I can't find the place where we decrease it after the peer with the max height disconnects.

Reply:
There is no notion of max peer height in this proposal (therefore it is never reset). Termination mechanism is different (based on number of pending requests, size of the peer set and termination timeout). I have added more comments in the latest version. MaxRequestPending is the maximum height for which request is made to some peer.